code/modules/reagents/chemistry/machinery/reagentgrinder.dm

```diff
@@ -232,73 +233,96 @@
 	return FALSE
 
 /obj/machinery/reagentgrinder/attack_hand(mob/user)
-	user.set_machine(src)
-	interact(user)
-
-/obj/machinery/reagentgrinder/interact(mob/user) // The microwave Menu
-	var/is_chamber_empty = 0
-	var/is_beaker_ready = 0
-	var/processing_chamber = ""
-	var/beaker_contents = ""
-	var/dat = ""
-
-	if(!operating)
-		for (var/obj/item/O in holdingitems)
-			processing_chamber += "\A [html_encode(O.name)]<BR>"
-
-		if(!processing_chamber)
-			is_chamber_empty = 1
-			processing_chamber = "Nothing."
-		if(!beaker)
-			beaker_contents = "<B>No beaker attached.</B><br>"
+	ui_interact(user)
+
+/obj/machinery/reagentgrinder/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "ReagentGrinder",  name, 400, 500, master_ui, state)
+		ui.open()
+
+/obj/machinery/reagentgrinder/ui_data(mob/user)
+	var/list/data = list()
+	data["name"] = name
+	data["operating"] = operating
+	data["inactive"] = length(holdingitems) == 0 ? TRUE : FALSE
+	data["limit"] = limit
+	data["count"] = length(holdingitems)
+	data["beaker_loaded"] = beaker ? TRUE : FALSE
+	data["beaker_current_volume"] = beaker ? beaker.reagents.total_volume : null
+	data["beaker_max_volume"] = beaker ? beaker.reagents.maximum_volume : null
+	var/list/beakerContents = list()
+	if(beaker)
+		for(var/datum/reagent/R in beaker.reagents.reagent_list)
+			beakerContents.Add(list(list("name" = R.name, "volume" = R.volume))) // list in a list because Byond merges the first list...
+	data["beaker_contents"] = beakerContents
+
+
+	var/list/items_counts = list()
+	var/list/name_overrides = list()
```

Comment:
imo this would be great if this was switched over to some kind of generalized proc for both this and kitchen machinery, but this is fine